### DISCUSSION TITLE
Large simulation with water spray runs out of virtual memory

### DISCUSSION THREAD

**Discussion by FireResearch-BK** [firemodels/fds#11940] [Q&A]
We have a simulation with the following basic setup:

- 32 meshes, cell size 0.2 m x 0.2 m x 0.2 each (9,442,274 cells in total; domain size ~300 x 20 x 20 m)
- Fire with 1000 kW/m², redefined HRR ramp
- Complex stochiometry with a defined soot yield creating quite a bit of smoke
- 30 water spray nozzles, active within 5 of the 32 meshes (1 x 3 nozzles, 1 x 6 nozzles, 3 x 7 nozzles; not directly spraying onto the fire)
- Control logic to activate all nozzles at 300 seconds
- 20 2D slices through the domain, WRITE_XYZ=T

Other, relevant, parameters:
- Particle: AGE=60.0, SAMPLING_FACTOR=20
- Spray model: PARTICLES_PER_SECOND=10000
- Simulation: MINIMUM_ZONE_VOLUME=0.07, SUSPEND_PRESSURE_ITERATIONS=T, VELOCITY_TOLERANCE=0.005, MAX_PRESSURE_ITERATIONS=50, WRITE_XYZ=.TRUE.
- (MAXIMUM_PARTICLES is at default)

The simulation runs on a virtual machine in a cloud server, with the following allocated hardware:
- Intel Xeon CPU with 32 physical cores (64 CPUs with HT)
- 124 GB physical RAM
- 200 GB of SSD space

The virtual machine runs Windows 10 Pro 64 bit. We set Windows's page file to a fixed size of 100 GB.

The FDS version in use is 6.8.0 and the case runs with MPI.

The simulation crashes after ~390 seconds due to running out of virtual memory:
`forrtl: severe(41): insufficient virtual memory`

HRR, in the simulation, at that point, was ~160 kW.

Open Hardware Monitor shows that just ~80 % of physical RAM was in use at the time of the crash.

The basic simulation setup is fixed any may not be changed (except for the number of meshes maybe).

Which of the parameters above (or which other parameters) can be tweaked to decrease (virtual) memory usage?

P.S: 
Starting the case from a restart point drives RAM usage down to 60% again. We suspect either a memory leak that got cleaned up by the restart or something like particles that went missing. Or maybe it's normal.

**Comment by drjfloyd** [comment]
Does sound like a memory leak somewhere. There was a particle bug that was recently fixed for culling particles when MAXIMUM_PARTICLES is hit. You may want to try the current test version of FDS. 

With 30 nozzles operating at once you are going to inject a lot of particles. You may bumping up against MAXIMUM_PARTICLES. Particle handling is a bit different for a particle killed by AGE or MINIMUM_DIAMETER than it is for removing the oldest particle when MAXIMUM_PARTICLES is hit, so another thing to consider is do you need AGE=60? If you don't care about what happens to particles once they have reached the floor, reducing AGE may let particles get culled for AGE before they are culled for MAXIMUM_PARTICLES.

**Comment by FireResearch-BK** [comment]
This is the memory usage up to the crash (left) and after the restart (right):
![grafik](https://github.com/firemodels/fds/assets/114649229/a6332499-9816-4390-a775-e77177dee7f6)

Memory consumption has remained steady since then and there seemed to be no loss of particle data due to the restart.
The simulation has so far passed the time of the crash (currently at ~405 s).


> Does sound like a memory leak somewhere. There was a particle bug that was recently fixed for culling particles when MAXIMUM_PARTICLES is hit. You may want to try the current test version of FDS.

It may be too early for this conclusion (considering there only have been a few time steps), but the lack of an increase in memory usage after the restart so far does not point to an issue with culling. The nozzles are still actively releasing particles into the domain (and will do so until T_end of the simulation).
We might give a nightly build a try in case we have to start over from the beginning due to an irrecoverable crash.

> With 30 nozzles operating at once you are going to inject a lot of particles. You may bumping up against MAXIMUM_PARTICLES. Particle handling is a bit different for a particle killed by AGE or MINIMUM_DIAMETER than it is for removing the oldest particle when MAXIMUM_PARTICLES is hit, so another thing to consider is do you need AGE=60? If you don't care about what happens to particles once they have reached the floor, reducing AGE may let particles get culled for AGE before they are culled for MAXIMUM_PARTICLES.

Decreasing AGE is kind of a last resort thing as we have a very long domain and also need the interaction of water with smoke near its end (away from the nozzles). So we wish to keep the particles in non-saturated meshes from being culled for as long as possible.

**Comment by IFAB-AS** [comment]
The restart did also crash at approx. 440 s (RAM usage 62%) and the version 6.8.0-332 did crash at a similar time as 6.8.0.
I was able to solve the problem for us by setting the AGE=15. In the test with one nozzle, we had the floor set as the domain boundary (solid) and the particles were removed from the domain. In the full model, there were several OBSTs set as the floor level and the particles were piling up.
With AGE=15, no particles were culled due to the maximum amount of particles per mesh (1.000.000) and the RAM usage was steady with spikes at the RESTART save points. If we use higher values for AGE, the RAM usage increases over time or the crash occurs even when there is plenty of RAM left. I assume, there is still some error in the storage/index settings for the culled particles in the MAXIMUM_PARTICLES routine, see https://github.com/firemodels/fds/issues/11897.
